libsolidity/ast/Types.cpp

```diff
@@ -2350,8 +2342,10 @@ u256 StructType::memoryDataSize() const
 	return size;
 }
 
-bigint StructType::storageSizeUpperBound() const
+bigint StructType::sizeUpperBound(DataLocation _loc) const
 {
+	if (_loc != DataLocation::Storage)
+		return 1;
```

Comment:
Many changes by this PR are more mechanical than semantic, trying to be close to the previous implementation. If there were no recursive walk before, there is still no one.

But that should be addressed, for sure. I am just not very familiar with internal storage/memory/calldata organization, so would like to address it later, when we are fine with the overall approach.

Reply:
Oh, that clarifies it a bit. I thought this PR was meant change stuff a bit more.

In that case I think that you should put `solUnimplemented()` here and add a special case for memory/calldata structs in `validForLocation()` to avoid calling it. Before your changes there were no size checks for such structs anyway so I think it's better to leave it at that than to add something that's potentially wrong.

I'm not sure what it should actually return in this case though. 1 slot would be the size of the head (i.e. reference) but I think that in `validForLocation()` we're more interested about the size of the struct as a whole, not just its head.

Reply:
I did similar thing for all data locations. (Not sure if it is quite accurate though.)

Reply:
Great. But in that case please also add a test that triggers this case :)
I think that something like `largeTypes/storage_parameter.sol` but with memory and calldata parameter should be enough.